components/resolved_ts/src/advance.rs

```diff
@@ -135,129 +140,144 @@ impl<E: KvEngine> AdvanceTsWorker<E> {
+                    tikv_clients,
+                    min_ts,
+                )
+                .await
             }
-        });
-        let resps = futures::future::join_all(stores).await;
-        resps
-            .into_iter()
-            .filter_map(|resp| match resp {
-                Ok(resp) => Some(resp),
-                Err(e) => {
-                    debug!("resolved-ts check leader error"; "err" =>?e);
-                    None
-                }
-            })
-            .map(|(store_id, resp)| {
-                resp.regions
-                    .into_iter()
-                    .map(move |region_id| (store_id, region_id))
-            })
-            .flatten()
-            .for_each(|(store_id, region_id)| {
+            .boxed()
+        })
+        .collect();
+    for _ in 0..store_count {
+        // Use `select_all` to avoid the process getting blocked when some TiKVs were down.
+        let (res, _, remains) = select_all(stores).await;
+        stores = remains;
+        if let Ok((store_id, resp)) = res {
+            for region_id in resp.regions {
                 resp_map.entry(region_id).or_default().push(store_id);
-            });
-        resp_map
-            .into_iter()
-            .filter_map(|(region_id, stores)| {
-                if region_has_quorum(&region_map[&region_id], &stores) {
-                    Some(region_id)
-                } else {
-                    debug!(
-                        "resolved-ts cannot get quorum for resolved ts";
-                        "region_id" => region_id,
-                        "stores" => ?stores,
-                        "region" => ?&region_map[&region_id]
-                    );
-                    None
+                if region_has_quorum(&region_map[&region_id], &resp_map[&region_id]) {
+                    valid_regions.insert(region_id);
                 }
-            })
-            .collect()
+            }
+        }
+        // Return early if all regions had already got quorum.
+        if valid_regions.len() == regions.len() {
+            // break here because all regions have quorum,
+            // so there is no need waiting for other stores to respond.
+            break;
+        }
+    }
+    valid_regions.into_iter().collect()
+}
+
+async fn send_check_leader_request(
+    store_id: u64,
+    regions: Vec<LeaderInfo>,
+    pd_client: Arc<dyn PdClient>,
+    security_mgr: Arc<SecurityManager>,
+    env: Arc<Environment>,
+    tikv_clients: Arc<Mutex<HashMap<u64, TikvClient>>>,
+    min_ts: TimeStamp,
+) -> Result<(u64, CheckLeaderResponse)> {
+    if tikv_clients.lock().unwrap().get(&store_id).is_none() {
+        let store = box_try!(pd_client.get_store_async(store_id).await);
+        let cb = ChannelBuilder::new(env.clone());
+        let channel = security_mgr.connect(cb, &store.address);
+        tikv_clients
+            .lock()
+            .unwrap()
+            .insert(store_id, TikvClient::new(channel));
```

Comment:
When do we release these connections?

Reply:
There is an issue https://github.com/tikv/tikv/issues/11400 about it, I think we can fix it in another PR.

Reply:
Fixed